### PR TITLE
feat(.github/actions): add `has-auto-releasable-commits-v1` action

### PR DESCRIPTION
**Pull request by Zidious**
This PR introduces a new GHA that checks if the list of commits (produced by `generate-commit-list-v1`) and whether we are version locked to axe-core or not if we should continue releasing or not. 

Coverage: https://github.com/dequelabs/axe-api-team-public/actions/runs/6177733830/job/16769523742?pr=26#step:5:15

Closes: https://github.com/dequelabs/axe-api-team/issues/373

### CODE REVIEW

File: .github/actions/has-auto-releasable-commits-v1/src/shouldAutoRelease.test.ts [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+              assert.isFalse(result)
+            })
+          })
+        })
+      })
+
+      describe('and there are breaking changes', () => {
+        it('should return false', () => {
+          const result = shouldAutoRelease({
```
Comment (straker) [inline .github/actions/has-auto-releasable-commits-v1/src/shouldAutoRelease.test.ts:172]: Should have tests for both `fix!` and `feat!`

File: .github/actions/has-auto-releasable-commits-v1/src/shouldAutoRelease.test.ts [hunk 2]
```diff
@@ -0,0 +1,227 @@
+import 'mocha'
+import { assert } from 'chai'
+import shouldAutoRelease from './shouldAutoRelease'
+
+describe('shouldAutoRelease', () => {
```
Comment (straker) [inline .github/actions/has-auto-releasable-commits-v1/src/shouldAutoRelease.test.ts:5]: This testing suite might be able to be simplified / more readable by using a test suite matrix with the different input types and expected output. Something like:

```js
const testMatrix = [
  {
    // not versioned locked with patch change
    versionLocked: false,
    commitType: 'fix',
    expected: true
  },
  {
    // version locked with patch change and axe-core patch
    versionLocked: true,
    axeCoreType: 'fix',
    commitType: 'fix',
    expected: true
  },
  {
    // no commits
    versionLocked: true,
    expected: false
  }
]
```

File: .github/actions/has-auto-releasable-commits-v1/src/run.test.ts [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+      assert.isTrue(setFailed.calledOnce)
+      assert.isTrue(
+        setFailed.calledWith(
+          'Invalid value for version-locked: foo. Must be true or false'
+        )
+      )
+    })
+  })
+
+  describe('when version-locked input is false', () => {
```
Comment (straker) [inline .github/actions/has-auto-releasable-commits-v1/src/run.test.ts:84]: See comment about using a testing matrix (as well as taking into account `fix`, `feat`, `fix!` and `feat!`)

**Review by straker — DISMISSED** [review]

File: .github/actions/has-auto-releasable-commits-v1/src/shouldAutoRelease.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    // perf: we don't need to continue if we have breaking changes or major/minor changes for axe-core
+    if (hasBreakingChanges || hasMajorOrMinorChangesForAxeCore) {
+      break
+    }
+
+    if (type === 'feat!') {
+      hasBreakingChanges = true
+    }
+
+    // This seems brittle if the update axe action changes the title this will break >.>
```
[the remark this answers is not in the document]
Comment (Zidious) [reply .github/actions/has-auto-releasable-commits-v1/src/shouldAutoRelease.ts:35]: https://github.com/dequelabs/axe-api-team-public/issues/30

File: .github/actions/has-auto-releasable-commits-v1/src/types.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import type core from '@actions/core'
+
+export type Core = Pick<
+  typeof core,
+  'getInput' | 'setOutput' | 'info' | 'setFailed'
+>
+
+export type CommitList = {
```
[the remark this answers is not in the document]
Comment (Zidious) [reply .github/actions/has-auto-releasable-commits-v1/src/types.ts:8]: It's here so I can do the work until the commits PR lands, I'll not merge this until I can pull in these specific types

File: .github/actions/has-auto-releasable-commits-v1/src/shouldAutoRelease.test.ts (hunk 2 above)
Comment (Zidious) [reply .github/actions/has-auto-releasable-commits-v1/src/shouldAutoRelease.test.ts:5]: I tried doing this, it becomes quite annoying when we start nesting cases, instead I've removed the unnecessary commit objects into: 

`createCommit` and `createAxeCoreCommit`

```ts
describe('and version is locked', () => {
      describe('and there are no breaking changes', () => {
        describe('and there are no major or minor changes for axe-core', () => {
          describe('and there is a feat commit', () => {
            it('should return true', () => {
              const result = shouldAutoRelease({
                commitList: [createCommit('feat')],
                isVersionLocked: true
              })

              assert.isTrue(result)
            })
          })

          describe('and there is a fix commit', () => {
            it('should return true', () => {
              const result = shouldAutoRelease({
                commitList: [createCommit('fix')],
                isVersionLocked: true
              })

              assert.isTrue(result)
            })
          })
        })
      })
})
```

Now should be easier to read the test case as:

```
When there are commits, and the version is locked, and there are no breaking changes, and there are no major or minor releases for axe-core, and there is a feat commit, returns true

// and so on
```

**Review by michael-siek — DISMISSED** [review]

File: .github/actions/has-auto-releasable-commits-v1/package.json [hunk 6]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "has-auto-releasable-commits-v1",
+  "version": "1.0.0",
+  "description": "A GitHub Action to determine if a parent action should continue or stop running",
```
Comment (straker) [inline .github/actions/has-auto-releasable-commits-v1/package.json:4]:
```suggestion
  "description": "A GitHub Action to determine if the repository has commits that should be automatically released",
```

File: .github/actions/has-auto-releasable-commits-v1/package.json [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  "bugs": {
+    "url": "https://github.com/dequelabs/axe-api-team-public/issues"
+  },
+  "homepage": "https://github.com/dequelabs/axe-api-team-public#readme",
+  "license": "UNLICENSED",
```
Comment (straker) [inline .github/actions/has-auto-releasable-commits-v1/package.json:9]:
```suggestion
  "license": "MPL-2.0",
```

**Review by straker — CHANGES_REQUESTED** [review]

**Review by straker — CHANGES_REQUESTED** [review]

File: .github/actions/has-auto-releasable-commits-v1/src/run.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    const commitList: Array<CommitList> = JSON.parse(commits)
+    const isVersionLocked = versionLocked === 'true'
+    const shouldRelease = shouldAutoRelease({
+      commitList,
+      isVersionLocked
+    })
+
+    core.setOutput('should-release', shouldRelease)
```
Comment (straker) [inline .github/actions/has-auto-releasable-commits-v1/src/run.ts:27]: We should `core.info` the result of `shoudlRelease` to make debugging easier

```suggestion
    core.info('Set should-release output: ' + shouldRelase)
    core.setOutput('should-release', shouldRelease)
```

**Review by straker — CHANGES_REQUESTED** [review]

**Review by straker — APPROVED** [review]

**Pull request merged**